### DISCUSSION TITLE
¿How do you load multiple GLTF/GLB to model viewer and then "merge" them into a single one?

### DISCUSSION THREAD

**Discussion by LaloYanez** [google/model-viewer#2276] [Q&A] [answered]
Yes, it sounds like I want to use multiple models inside a single instance but its not the case. Let me explain a little.

I have recently stumbled upon a very nicely done car configurator built on model viewer. After exploring the options I noticed some things that felt off and took a peek at the code. First, the env map was not a sphere as usual but had a floor, it was possible to switch to different environments or none without loading a different model. The same thing was happening when different animations (doors opening, wheels rolling, airbags, etc. were not only playing separately but could also be active at the same time.

What I found is that indeed, there were 4 different gltf models in play, one for the dome that held the env map texture, the wheels, the inside and the outside of the car. I could not figure out how it was done by looking at obfuscated code but it was clear that somehow the gltf models were getting loaded and then merged in runtime. 

I could not find anything about this after several hours of research and all I know is that the models were using draco compression, not sure if that has anything to do with it or its just to make the models lighter.

Any ideas?

**Comment by elalish** [comment, marked as answer]
Can you share a link? We do not yet support this, but we plan to add this kind of functionality to our scene-graph API. It's possible that they hacked their way down into the Three.js layer to make all this work. Definitely possible, but not encouraged as it's likely to be fragile whenever we make a new release.

**Reply by LaloYanez** [reply]
https://www.vw.com.mx/app/virtual-studio/index.html

Also noticed that they can move the camera while inside the car with device orientation data when given permissions. That would also be a nice addition to model viewer 😄

**Reply by elalish** [reply]
Are you sure that's built on model-viewer? I think they rolled their own.

**Reply by LaloYanez** [reply]
It is, they set up the html with empty attributes that get loaded later on with json data. Its the most customized experience ive seen on model viewer, so much that it doesnt look like its model viewer 😄

I only noticed at first that was indeed model viewer because I was doing a similar thing and was already familiar with all the things you could do like camera orbit, target, fov, auto rotation and quick look AR session on iOS whith a usdz static model.

**Reply by elalish** [reply]
Oh wow, you're right! I just tried it in WebXR mode and sure enough. And yeah, that look around feature inside the car is really nice. We will definitely keep this use in mind as we build out the scene-graph API, to make this kind of experience easier to build.

**Reply by LaloYanez** [reply]
Impressive huh!

Whenever I have the time Ill try to crack that multi model mystery, could be useful for configurator type experiences.

**Reply by felipe-chamas** [reply]
I think the "viewer of models" is implemented with Threejs while the `model-viewer` is used empty only to load the AR experience. If you search through Sources tab on devTools the implementation is Canvas based and has more to do with pure Threejs implementation.

**Reply by felipe-chamas** [reply]
Viewer implemented in three:
![image](https://user-images.githubusercontent.com/49691924/114237644-c1f30380-9959-11eb-8d5f-9bc5fa3b55e6.png)
AR using model-viewer:
![image](https://user-images.githubusercontent.com/49691924/114237752-e5b64980-9959-11eb-8bf1-82124545674a.png)

**Reply by LaloYanez** [reply]
Ohh makes sense, its a nice concept, close to what elalish thought they did. Thanks to both for figuring it out!
